cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java

```diff
@@ -143,69 +144,87 @@ private static String httpServletRequestToString(HttpServletRequest request) {
-                                                   int step) {
+  public List<OperationFuture> getOrCreateUserTask(HttpServletRequest httpServletRequest, HttpServletResponse httpServletResponse,
+      Function<String, OperationFuture> function, int step, boolean asyncComplete) {
     UUID userTaskId = getUserTaskId(httpServletRequest);
-    List<OperationFuture> operationFutures = getFuturesByUserTaskId(userTaskId, httpServletRequest);
+    UserTaskInfo userTaskInfo = getUserTaskByUserTaskId(userTaskId, httpServletRequest);
 
-    if (operationFutures != null) {
+    if (!asyncComplete && userTaskInfo != null) {
+      throw new IllegalArgumentException(
+          String.format("Sync operation should not have %s", USER_TASK_HEADER_NAME));
+    }
+
+    if (userTaskInfo != null) {
       LOG.info("Fetch an existing UserTask {}", userTaskId);
       httpServletResponse.setHeader(USER_TASK_HEADER_NAME, userTaskId.toString());
-      if (step < operationFutures.size()) {
-        return operationFutures;
-      } else if (step == operationFutures.size()) {
+      if (step < userTaskInfo.futures().size()) {
+        return userTaskInfo.futures();
+      } else if (step == userTaskInfo.futures().size()) {
         LOG.info("Add a new future to existing UserTask {}", userTaskId);
-        return insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
+        return insertFuturesByUserTaskId(userTaskId, function, httpServletRequest).futures();
       } else {
         throw new IllegalArgumentException(
-            String.format("There are %d steps in the session. Cannot add step %d.", operationFutures.size(), step));
+            String.format("There are %d steps in the session. Cannot add step %d.", userTaskInfo.futures().size(), step));
       }
     } else {
-      if (httpServletRequest.getHeader(USER_TASK_HEADER_NAME) != null) {
-        // request provides user_task_header that user tasks doesn't exist then just throw exception
-        String userTaskIdFromRequest = httpServletRequest.getHeader(USER_TASK_HEADER_NAME);
-        throw new IllegalArgumentException(
-            String.format("UserTask %s is an invalid %s", userTaskIdFromRequest, USER_TASK_HEADER_NAME));
-      }
-
+      ensureUserTaskHeaderNotPresent(USER_TASK_HEADER_NAME, httpServletRequest);
       if (step != 0) {
         throw new IllegalArgumentException(
             String.format("There are no step in the session. Cannot add step %d.", step));
       }
-
-      SessionKey sessionKey = new SessionKey(httpServletRequest);
       userTaskId = _uuidGenerator.randomUUID();
-      LOG.info("Create a new UserTask {} with SessionKey {}", userTaskId, sessionKey);
-      operationFutures = insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
-
-      synchronized (_sessionKeyToUserTaskIdMap) {
-        _sessionKeyToUserTaskIdMap.put(sessionKey, userTaskId);
+      userTaskInfo = insertFuturesByUserTaskId(userTaskId, function, httpServletRequest);
```

Comment:
Note in method `insertFuturesByUserTaskId `, there is [limit](https://github.com/linkedin/cruise-control/blob/f7e15c4aa6c784129dfd196355c8454e89ede4f3/cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/servlet/UserTaskManager.java#L326) on number of active tasks(currently is 5).  

Consider the scenario that there some requests get hang in execution and use up the slot,  then user cannot send `user_tasks` request to check because this task insertion will fail. I think we probably need to take special care to `user_tasks` request here.